### PR TITLE
[V2] Remote improvements

### PR DESCRIPTION
**Pull request by lmr**
This contains a number of changes made to support my AWS cluster tests.

Those changes basically refactor a number of things in the remote class, enable multiple instances to be executed on the same test, and work around a Paramiko bug [2].

This is a rebase against the latest master. I have a new SSH library, but I imagine that will take more time to get in. Meanwhile, I have other tests, and a plugin that depend on those changes.

[1] Possibly Paramiko as well (the keys handling bug is specially annoying). I'm seriously thinking about reusing the autotest SSH host classes and have something that works with raw ssh and paramiko interchangeably.
[2] https://github.com/paramiko/paramiko/issues/387

### CODE REVIEW

**Comment by clebergnu** [comment]
Hi @lmr, thanks for the v2! I have gone through the code and have not found any big issues (besides the CI issues which I'll take a look at).  Still, functionality wise, it's bringing be strange results.  I get the job quietly "aborted" right after authenticating to the remote end:

```
$ ./scripts/avocado run passtest --remote-hostname 127.0.0.1 --remote-port=2222 --remote-username root --remote-password redhat 
LOGIN      : root@127.0.0.1:2222 (TIMEOUT: 60 seconds)
$
```

Since we're approaching a release targeted at stability, having those improvements is very welcome.  I'll spend some time looking at those issues.

Thanks again!

**Comment by lmr** [comment]
@clebergnu yeah I'm tracking that problem, sorry about that.

**Comment by clebergnu** [comment]
@lmr don't you apologize! I'll also try to get back on this after the sun rises around here.

**Comment by clebergnu** [comment]
Functional test is sound.  I'll briefly go over the code now.

File: avocado/core/remoter.py [hunk 1]
```diff
@@ -285,6 +303,7 @@ def send_files(self, local_path, remote_path):
                                            remote_path, hosts=[self.hostname])
         return result_dict[self.hostname]
 
+    @_update_fabric_env
```
Comment (clebergnu) [inline avocado/core/remoter.py]: Right. Looks OK to me.

File: avocado/core/remoter.py [hunk 2]
```diff
@@ -178,41 +275,23 @@ def makedir(self, remote_path):
...
         """
-        try:
-            fabric.operations.get(remote_path,
-                                  local_path)
-        except ValueError:
-            return False
-        return True
+        result_dict = fabric.tasks.execute(receive_files, local_path,
+                                           remote_path, hosts=[self.hostname])
+        return result_dict[self.hostname]
```
Comment (clebergnu) [inline avocado/core/remoter.py]: OK, I see the goal and I like it.

File: selftests/unit/test_vm.py [hunk 3]
```diff
@@ -35,6 +35,7 @@ def setUp(self):
                         vm_hostname='hostname',
                         vm_port=22,
                         vm_password='password',
+                        vm_key_file=None,
```
Comment (clebergnu) [inline selftests/unit/test_vm.py]: OK, looks good.

File: avocado/core/remoter.py [hunk 4]
```diff
@@ -101,8 +101,11 @@ def run(self, command, ignore_status=False, timeout=60):
         Run a remote command.
 
         :param command: the command string to execute.
+        :param ignore_status: Whether to not raise exceptions in case the
+            command's return code is different than zero.
+        :param timeout: Maximum time allowed for the command to return.
 
-        :return: the result of the remote program's output.
+        :return: the result of the remote program's execution.
```
Comment (clebergnu) [inline avocado/core/remoter.py]: OK.

File: avocado/core/remoter.py [hunk 5]
```diff
@@ -123,7 +123,7 @@ def _run(self, command, ignore_status=False, timeout=60):
         while True:
             try:
                 fabric_result = fabric.operations.run(command=command,
-                                                      quiet=self.quiet,
+                                                      quiet=quiet,
```
Comment (clebergnu) [inline avocado/core/remoter.py]: OK.

File: avocado/core/remoter.py [hunk 6]
```diff
@@ -83,23 +83,18 @@ def __init__(self, hostname, username=None, password=None,
...
-                                abort_on_prompts=True,
-                                abort_exception=ConnectionError,
-                                reject_unknown_hosts=reject_unknown_hosts,
-                                disable_known_hosts=disable_known_hosts)
-
-    @staticmethod
-    def _setup_environment(**kwargs):
-        """ Setup fabric environemnt """
-        fabric.api.env.update(kwargs)
+        fabric.api.env.update(host_string=hostname,
```
Comment (clebergnu) [inline avocado/core/remoter.py]: OK.

File: avocado/core/remoter.py [hunk 7]
```diff
@@ -187,6 +198,12 @@ def send_files(self, local_path, remote_path):
         return True
 
     def receive_files(self, local_path, remote_path):
+        result_dict = fabric.tasks.execute(self._receive_files, local_path,
+                                           remote_path, hosts=[self.hostname])
+        return result_dict[self.hostname]
+
+    @staticmethod
+    def _receive_files(self, local_path, remote_path):
```
Comment (clebergnu) [inline avocado/core/remoter.py]: OK.

File: avocado/core/remoter.py [hunk 8]
```diff
@@ -78,7 +88,7 @@ def run(command, ignore_status=False, quiet=True, timeout=60):
                                                   warn_only=True,
                                                   timeout=timeout)
             break
-        except fabric.network.NetworkError, details:
+        except (fabric.network.NetworkError, AuthenticationError), details:
```
Comment (clebergnu) [inline avocado/core/remoter.py]: Nitpick (Py3k compat): `except <klass> as details`

**Comment by clebergnu** [comment]
Merged without the last commit, which became unnecessary due to #1132

**Pull request closed without merging**